ReactiveCocoa/Swift/Signal.swift

```diff
@@ -139,22 +84,129 @@ public final class Signal<Value, Error: ErrorProtocol> {
 	///            the Signal itself.
 	@discardableResult
 	public func observe(_ observer: Observer) -> Disposable? {
+		if let disposable = stream.observe(observer) {
+			return disposable
+		} else {
+			observer.sendInterrupted()
+			return nil
+		}
+	}
+}
+
+private final class EventStream<Value, Error: ErrorProtocol> {
+	/// The optional disposable returned by the signal generator.
+	/// It would be disposed of when the event stream is disposed of.
+	private var generatorDisposable: Disposable?
+
+	/// An atomic bag of observers to this event stream.
+	private let atomicObservers: Atomic<Bag<Signal<Value, Error>.Observer>?>
```

Comment:
I realize that you didn't name it, but this name is a bit silly. It should just be called `observers`. The atomicity is noted in the type.